tasklog/log.go

```diff
@@ -39,8 +50,9 @@ type Logger struct {
 }
 
 // NewLogger retuns a new *Logger instance that logs to "sink" and uses the
-// current terminal width as the width of the line.
-func NewLogger(sink io.Writer) *Logger {
+// current terminal width as the width of the line. Will log progress status if
+// stdout is a terminal or if forceProgress is true
+func NewLogger(sink io.Writer, forceProgress bool) *Logger {
```

Comment:
I think that it may be good to adopt the same pattern of [option functions][1] here, too, even though there is only one option that we'd be supporting. I suggest this, because I think the "extra baggage" of option functions is worth the tradeoff to remove a boolean argument. Perhaps:

```go
type optionFn func(l *Logger)

var (
        forced optionFn = func(l *Logger) {
                l.forceProgress = true
        }
)

// NB: this prevents mutating the underlying variable, which _can_ be changed at runtime.
func Forced() optionFn { return forced }

func NewLogger(sink io.Writer, fns ...optionFn) {
        l := &Logger{...}
        for _, fn := range fns {
                fn(l)
        }

        // ...
}
```

Then to call this (with forcing) you would instead say:

```go
l := tasklog.NewLogger(os.Stderr, tasklog.Forced())
```

I think that you would want a helper function that calls that ☝️, since it removes the convenience of simply passing along `cfg.ForceProgress()`.

[1]: https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

Reply:
Agree. Nice pattern; hadn't seen that before. Thanks for the pointer.

I'm not sure I understand the _"NB: this prevents mutating the underlying variable_" comment, though.